dsgrid/utils/dataset.py

```diff
@@ -142,6 +161,112 @@ def is_noop_mapping(records: DataFrame) -> bool:
     )
 
 
+def map_time_dimension_with_chronify_duckdb(
+    df: DataFrame,
+    value_column: str,
+    from_time_dim: TimeDimensionBaseConfig,
+    to_time_dim: TimeDimensionBaseConfig,
+) -> DataFrame:
+    """Create a time-mapped table with chronify and DuckDB.
+    All operations are performed in memory.
+    """
+    src_schema, dst_schema = _get_mapping_schemas(df, value_column, from_time_dim, to_time_dim)
+    store = chronify.Store.create_in_memory_db()
+    store.ingest_table(df.relation, src_schema, bypass_checks=True)
+    store.map_table_time_config(src_schema.name, dst_schema)
+    pandas_df = store.read_table(dst_schema.name)
+    store.drop_table(dst_schema.name)
+    return df.session.createDataFrame(pandas_df)
+
+
+def map_time_dimension_with_chronify_spark_hive(
+    df: DataFrame,
+    table_name: str,
+    value_column: str,
+    from_time_dim: TimeDimensionBaseConfig,
+    to_time_dim: TimeDimensionBaseConfig,
+) -> DataFrame:
+    """Create a time-mapped table with chronify and Spark and a Hive Metastore.
+    The source data must already be stored in the metastore.
+    Chronify will store the mapped table in the metastore.
+    """
+    src_schema, dst_schema = _get_mapping_schemas(
+        df, value_column, from_time_dim, to_time_dim, src_name=table_name
+    )
+    store = chronify.Store.create_new_hive_store(dsgrid.runtime_config.thrift_server_url)
+    with store.engine.begin() as conn:
+        # This bypasses checks because the table should already be valid.
+        store.schema_manager.add_schema(conn, src_schema)
+    try:
+        # TODO: What about wrap_time_allowed?
```

Comment:
If I recall, wrap_time_allowed = True allows wrapping time to be applied to unaligned time.
If dataset time becomes misaligned from project time after interval type adjustment, it is wrapped automatically.

Chronify is supporting the latter with roll_time_interval(). We haven't connected the pieces on wrap_time_allowed yet, which will be coming in from a data_adjustment_model. I think the index time for the industrial dataset may need that.

Reply:
@lixiliu will write an issue about remaining.